handler/MenuHandler.py

```diff
@@ -2,66 +2,170 @@
-		pass
+		for neighbour in neighbours:
+			self.__unicast(
+				AppData.get_peer_ip4(neighbour),
+				AppData.get_peer_ip6(neighbour),
+				AppData.get_peer_port(neighbour),
+				request)
 
 	def serve(self, choice: str) -> None:
 		""" Handle the peer request
-		Parameters:
-			request - the list containing the request parameters
-		Returns:
-			str - the response
+
+		:param choice: the choice to handle
+		:return: None
 		"""
 
 		if choice == "QUER":
+
 			# codice che manda il pkt sulla socket
+			pktid = str(uuid.uuid4().hex[:16].upper())
+			ip = '172.016.001.001|FC00:2001:db8a:a0b2:12f0:a13w:0001:0001'
+			port = '4000'
+
+			search = input('\nEnter the file name: ')
+
+			# aggiungere verifica lunghezza file name
+			if len(search) <= 0 or len(search) > 20:
+				print('File name must be between 1 and 20 chars long.\n')
+				return
+			elif len(search) < 20:
+				# codice per riempire la search con spazi se necessario
+				zfill = 20 - len(search)
+				search = search + (' ' * zfill)
+
+			request = choice+pktid+ip+port+search
+
+			# avvio il server di ricezione delle response, lo faccio prima del broadcast
+			# per evitare che i primi client che rispondono non riescano a connettersi
+			t = Thread(target=lambda: Server(int(port), SelfHandler()).run(True))
+			t.daemon = True
+			t.start()
 
-			# avvio il server di ricezione delle response
-			p = Process(target=lambda: Server(4000, SelfHandler()).run())
-			p.daemon = True
-			p.start()
+			self.__broadcast(request)
 
-			# avvio il timer
-			t = Timer(300, self.__kill_server, args=(p,))
-			t.start()
+			# grazie alla join possiamo aspettare la fine dell'esecuzione di un thread (il server che riceve tutte le risposte)
+			t.join()
 
-			input('\nFile search in progress, press enter to continue...\n')
+			files = AppData.get_peer_files()
 
-			if t.isAlive:
-				t.cancel()
-				self.__kill_server(p)
+			for count, file in enumerate(files, start=1):
+				print(f'{count}]', file)
 
-			choice = input('Please select a peer:')
+			while True:
+				index = input('Please select a file to download:')
+				try:
+					index = int(index)
+					if 1 <= index < len(files):
+						break
+					else:
+						print(f'Index chosen must be in the correct range: 1 - {len(files)}\n')
+				except ValueError:
+					print('Your choice must be a valid one. Number in range 1 - {len(files} expected\n')
 
-		elif choice == "NEAR":
-			pass
+			host = (AppData.get_peer_ip4(files[index]), AppData.get_peer_ip6(files[index]), AppData.get_peer_port(files[index]))
+			file_md5 = AppData.get_file_md5(files[index])
+			file_name = AppData.get_file_name(files[index])
 
-		elif choice == "RETR":
+			# preparo request per retr, faccio partire server in attesa download, invio request e attendo
+			request = 'RETR'+file_md5+file_name
 
-			if len(request) != 36:
-				return "Invalid request, usage is RETR<Filemd5>"
+			(sock, port) = self.__unicast(host, request)
+			Downloader.start(sock, port, AppData.get_file_download())
```

Comment:
Fede tu lanci il Downloader passandogli come filename quello memorizzato in AppData, ma non lo hai settato? 
Quando uno inserosce l'index del file che vuole scaricare tu devi settarlo come variabile condivisa